libs/platform/ebpf_ring_buffer.c

```diff
@@ -0,0 +1,261 @@
+        _ring_advance_producer_offset(ring, requested_length);
+        record->length = requested_length;
+    }
+    return record;
+}
+
+/**
+ * @brief Free any records in the ring buffer older than ebpf_get_released_epoch as discarded.
+ *
+ * @param[in] context Pointer to the ringbuffer.
+ */
+static void
+_ring_buffer_flush(void* context)
+{
+    ebpf_ring_buffer_t* ring = (ebpf_ring_buffer_t*)context;
+    ebpf_lock_state_t state = ebpf_lock_lock(&ring->lock);
+    while (_ring_get_producer_offset(ring) != _ring_get_consumer_offset(ring)) {
+        ebpf_ring_buffer_record_t* record = _ring_next_consumer_record(ring);
+        // Is this a discarded record?
+        if (record->reserved_epoch == 0) {
+            break;
+            // Is this a reserved record that hasn't expired?
+        } else if (record->reserved_epoch >= ebpf_get_released_epoch()) {
+            break;
+        }
+        _ring_advance_consumer_offset(ring, record->length);
+    }
+    ebpf_lock_unlock(&ring->lock, state);
+}
+
+ebpf_result_t
+ebpf_ring_buffer_create(_Outptr_ ebpf_ring_buffer_t** ring, size_t capacity)
+{
+    ebpf_result_t result;
+    ebpf_ring_buffer_t* local_ring_buffer = ebpf_epoch_allocate(sizeof(ebpf_ring_buffer_t));
+    if (!local_ring_buffer) {
+        result = EBPF_NO_MEMORY;
+        goto Done;
+    }
+
+    if ((capacity & ~(capacity - 1)) != capacity) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
+    local_ring_buffer->length = capacity;
+
+    local_ring_buffer->ring_descriptor = ebpf_map_ring(capacity);
+    if (!local_ring_buffer->ring_descriptor) {
+        result = EBPF_NO_MEMORY;
+        goto Done;
+    }
+    local_ring_buffer->shared_buffer = ebpf_ring_descriptor_get_base_address(local_ring_buffer->ring_descriptor);
+
+    local_ring_buffer->flush_work_item = ebpf_epoch_allocate_work_item(local_ring_buffer, _ring_buffer_flush);
+
+    *ring = local_ring_buffer;
+    local_ring_buffer = NULL;
+    result = EBPF_SUCCESS;
+
+Done:
+    ebpf_ring_buffer_destroy(local_ring_buffer);
+    local_ring_buffer = NULL;
+    return result;
+}
+
+void
+ebpf_ring_buffer_destroy(_Frees_ptr_opt_ ebpf_ring_buffer_t* ring)
+{
+    if (ring) {
+        ebpf_epoch_free_work_item(ring->flush_work_item);
+        ebpf_unmap_ring(ring->ring_descriptor);
+    }
+    ebpf_epoch_free(ring);
+}
+
+ebpf_result_t
+ebpf_ring_buffer_output(_In_ ebpf_ring_buffer_t* ring, _In_reads_bytes_(length) uint8_t* data, size_t length)
+{
+    ebpf_result_t result;
+    ebpf_lock_state_t state = ebpf_lock_lock(&ring->lock);
+    ebpf_ring_buffer_record_t* record = _ring_buffer_allocate_record(ring, length);
+
+    if (record == NULL) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
+    record->reserved_epoch = 0;
+    memcpy(record->data, data, length);
+    result = EBPF_SUCCESS;
+Done:
+    ebpf_lock_unlock(&ring->lock, state);
+    return result;
+}
+
+ebpf_result_t
```

Comment:
do you envision any errors here in the future?  If not, change to void so the caller doesn't need to test for error.

Reply:
Was originally going to have an option to block until ring is not empty, but decided to move that to the map layer.